go/libkb/sig_chain.go

```diff
@@ -665,7 +699,81 @@ func (sc *SigChain) VerifySigsAndComputeKeys(ctx context.Context, eldest keybase
 	// the id_table.  See LoadUser in user.go and
 	// https://github.com/keybase/go/issues/43
 
-	return
+	return cached, len(links), nil
+}
+
+func reverseListOfChainLinks(arr []ChainLinks) {
+	for i, j := 0, len(arr)-1; i < j; i, j = i+1, j-1 {
+		arr[i], arr[j] = arr[j], arr[i]
+	}
+}
+
+// VerifySigsAndComputeKeys iterates over all potentially all incarnations of the user, trying to compute
+// multiple subchains. It returns (bool, error), where bool is true if the load hit the cache, and false othewise.
+func (sc *SigChain) VerifySigsAndComputeKeys(ctx context.Context, eldest keybase1.KID, ckf *ComputedKeyFamily, loadAllSubchains bool) (bool, error) {
+	// First consume the currently active sigchain.
+	cached, numLinksConsumed, err := sc.verifySigsAndComputeKeysCurrent(ctx, eldest, ckf)
+	if !loadAllSubchains || err != nil || ckf.kf == nil {
+		return cached, err
+	}
+
+	allCached := cached
+
+	// ignore error here, since it shouldn't kill the overall load if historical subchains don't run
+	// correctly.
+	cached, _ = sc.verifySigsAndComputeKeysHistorical(ctx, numLinksConsumed, *ckf.kf)
+	if !cached {
+		allCached = false
+	}
+	return allCached, nil
+}
+
+func (sc *SigChain) verifySigsAndComputeKeysHistorical(ctx context.Context, numLinksConsumed int, kf KeyFamily) (allCached bool, err error) {
+
+	defer sc.G().CTrace(ctx, "verifySigsAndComputeKeysHistorical", func() error { return err })()
+	allLinks := sc.chainLinks
+	var cached bool
+
+	var prevSubchains []ChainLinks
+	sc.G().Log.CDebugf(ctx, "Continuing chain iteration into history: %d links total, but only %d consumed", len(allLinks), numLinksConsumed)
+
+	for {
+		allLinks = allLinks[0 : len(allLinks)-numLinksConsumed]
+		if len(allLinks) == 0 {
+			sc.G().Log.CDebugf(ctx, "Ending iteration through previous subchains; no futher links")
+			break
+		}
+
+		i := len(allLinks) - 1
+		eldest := allLinks[i].ToEldestKID()
+		if eldest.IsNil() {
+			sc.G().Log.CDebugf(ctx, "Ending iteration through previous subchains; saw a nil eldest (@%d)", i)
+			break
+		}
+		sc.G().Log.CDebugf(ctx, "Examining subchain that ends at %d with eldest %s", i, eldest)
+
+		var links ChainLinks
+		links, err = backtrackUntilNextEldest(allLinks, eldest)
+		if err != nil {
+			sc.G().Log.CInfof(ctx, "Error backtracking all links from %d: %s", i, err)
+			break
+		}
+
+		cached, _, err = sc.verifySubchain(ctx, kf, links)
+		if err != nil {
+			sc.G().Log.CInfof(ctx, "Error verifying subchain from %d: %s", i, err)
+			break
```

Comment:
We probably don't have enough information to compute this for legacy chains, since there could have been a whole bunch of eldest swaps without any sigs.